r/R/dataset.R

```diff
@@ -163,30 +169,59 @@ DatasetFactory$create <- function(sources) {
 #' - `$Finish(schema)`: Returns a `Source`
 #' @rdname Source
 #' @name Source
-#' @seealso [Dataset] for what do do with a `Source`
+#' @seealso [Dataset] for what to do with a `Source`
 #' @export
 Source <- R6Class("Source", inherit = Object,
   active = list(
     #' @description
     #' Return the Source's `Schema`
     schema = function() {
       shared_ptr(Schema, dataset___Source__schema(self))
-    }
+    },
+    #' @description
+    #' Return the Source's type.
+    type = function() dataset___Source__type_name(self)
+  )
+)
+
+#' @name FileSystemSource
+#' @rdname Source
+#' @export
+FileSystemSource <- R6Class("FileSystemSource", inherit = Source,
+  active = list(
+    #' @description
+    #' Return the files contained in this `Source`
+    files = function() dataset___FSSource__files(self),
+    #' @description
+    #' Return the format of files in this `Source`
+    format = function() FileFormat$create(dataset___FSSource__format(self))
   )
 )
 
+Source$create <- function(x) {
```

Comment:
We don't use the `$create` methods for taking pointers and wrapping them in their class, generally. `create` is for R user facing input. Maybe the `..dispatch` pattern is a better fit.

Reply:
Moved to dispatch.